conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosGovernanceOperations.scala

```diff
@@ -0,0 +1,203 @@
+  )
+
+  //wrappings to make the distinction between two different counts
+  private case class BallotsPerCycle(counts: Voting.BallotCounts) extends AnyVal
+  private case class BallotsPerLevel(counts: Voting.BallotCounts) extends AnyVal
+
+  /** Collects rolls of votes counted during ballots
+    *
+    * @param yay how many yays
+    * @param nay how many nays
+    * @param pass how many passes
+    */
+  case class VoteRollsCounts(yay: Int, nay: Int, pass: Int)
+
+  /** We take basic governance data and the blocks storing that
+    * with the goal of deriving more detailed aggregate data on
+    * the voting process.
+    * The computation needs to access the database to fetch essential
+    * ballot data
+    *
+    * SAFETY NOTE: make sure that the blocks passed to this method have
+    *              have a proper metadata value, different from the genesis one
+    *
+    * @param db needed to fetch previously stored voting data
+    * @param proposalsBlocks blocks we want to have governance data for, associated to the protocol of the proposal
+    * @param proposalsBallots votes cast on the blocks of interest,
+    *                         the keys should be a superset of the `proposalsBlocks` keys,
+    *                         so that the value is available for each of those blocks.
+    * @param levelsRolls listings of baker rolls involved, indexed per level, we expect all levels for the
+    *                         `proposalsBlocks` to be available, plus the level immediately before this batch.
+    *                          ideally this would be a `Level => List[Voting.BakerRolls]` (i.e. a total function)
+    * @param ec needed to compose concurrent operations
+    * @return aggregated data
+    */
+  def aggregateData(db: Database)(
+      proposalsBlocks: Map[Block, ProtocolId],
+      proposalsBallots: Map[Block, List[Voting.Ballot]],
+      levelsRolls: Map[Int, List[Voting.BakerRolls]]
+  )(
+      implicit ec: ExecutionContext
+  ): Future[List[GovernanceAggregate]] = {
+
+    //local functions, should simplify reading the calling code
+    def countBallotsPerLevel(block: Block) =
+      TezosDatabaseOperations.getBallotOperationsForLevel(block.data.header.level)
+
+    // as stated, this is a runtime failure if the block contains the wrong metadata
+    def countBallotsPerCycle(block: Block) = block.data.metadata match {
+      case BlockHeaderMetadata(_, _, _, _, _, level) =>
+        TezosDatabaseOperations.getBallotOperationsForCycle(level.cycle)
+    }
+
+    // as stated, this is a runtime failure if the block contains the wrong metadata
+    def proposalOperationsHashes(block: Block) = block.data.metadata match {
+      case BlockHeaderMetadata(_, _, _, _, _, level) =>
+        TezosDatabaseOperations.getProposalOperationHashesByCycle(level.cycle)
+    }
+
+    logger.info("Searching for governance data in voting period...")
+
+    //main algorithm
+    val ballotCountsResult =
```

Comment:
We can save time here as well. `proposalsBlocks` contains data from three governance periods. However, ballots are only cast during the testing_vote and promotion_vote periods. If we filter out the blocks for the testing period, we will save many database calls.

Reply:
will do